crates/net/network/src/transactions.rs

```diff
@@ -524,47 +545,204 @@ where
+                "some hashes in announcement from peer didn't fit in `GetPooledTransactions` request, buffering surplus hashes"
+            );
+            self.transaction_fetcher.buffer_hashes(surplus_hashes, Some(peer_id));
+        }
+
+        trace!(target: "net::tx",
+            peer_id=abbrev_hex(peer_id.as_ref()),
+            hashes=abbrev_hex_hash_list(&hashes),
+            "sending hashes in `GetPooledTransactions` request to peer's session"
+        );
+
+        // request the missing transactions
+        let metrics = &self.metrics;
+        if let Some(failed_to_request_hashes) =
+            self.transaction_fetcher.request_transactions_from_peer(hashes, peer, || {
+                metrics.egress_peer_channel_full.increment(1)
+            })
+        {
+            debug!(target: "net::tx",
+                peer_id=abbrev_hex(peer_id.as_ref()),
+                hashes=%abbrev_hex_hash_list(&failed_to_request_hashes),
+                "sending `GetPooledTransactions` request to peer's session failed, buffering hashes"
+            );
+            self.transaction_fetcher.buffer_hashes(failed_to_request_hashes, Some(peer_id));
+            return
         }
 
         if num_already_seen > 0 {
+            self.metrics.messages_with_already_seen_hashes.increment(1);
+            trace!(target: "net::tx", num_hashes=%num_already_seen, ?peer_id, client=?peer.client_version, "Peer sent already seen hashes");
             self.report_already_seen(peer_id);
         }
     }
 
+    // Requests buffered hashes for which an idle peer exists.
+    fn request_buffered_hashes(&mut self) {
+        loop {
+            let mut hashes = vec![];
+            let Some(peer_id) = self.pop_any_idle_peer(&mut hashes) else {
+                return;
+            };
+
+            debug_assert!(
+                self.peers.contains_key(&peer_id),
+                "broken invariant `peers` and `transaction-fetcher`"
+            );
+
+            // fill the request with other buffered hashes that have been announced by the peer
+            if let Some(peer) = self.peers.get(&peer_id) {
+                let is_valid_for_version_eth68 = peer.version == EthVersion::Eth68;
+                self.transaction_fetcher.fill_request_for_peer(
+                    &mut hashes,
+                    peer_id,
+                    is_valid_for_version_eth68,
+                );
+
+                trace!(
+                    target: "net::tx",
+                    peer_id=abbrev_hex(peer_id.as_ref()),
+                    hashes=abbrev_hex_hash_list(&hashes),
+                    "requesting buffered hashes from idle peer"
+                );
+
+                // request the buffered missing transactions
+                let metrics = &self.metrics;
+                if let Some(failed_to_request_hashes) =
+                    self.transaction_fetcher.request_transactions_from_peer(hashes, peer, || {
```

Comment:
I see, here the peer should always have sufficient capacity for this request because we ensured above that it is `idle`

Reply:
yepp :)